lib/src/git.rs

```diff
@@ -315,69 +301,22 @@ fn export_changes(
     git_repo: &git2::Repository,
 ) -> Result<Vec<String>, GitExportError> {
-    upgrade_old_export_state(mut_repo.base_repo());
-
-    let last_export_path = mut_repo.base_repo().repo_path().join("git_export_view");
-    let last_export_store_view =
-        if let Ok(mut last_export_file) = OpenOptions::new().read(true).open(&last_export_path) {
-            let thrift_view = simple_op_store::read_thrift(&mut last_export_file)
-                .map_err(|err| GitExportError::ReadStateError(err.to_string()))?;
-            op_store::View::from(&thrift_view)
-        } else {
-            op_store::View::default()
-        };
-    let last_export_view = View::new(last_export_store_view);
-    let (new_export_store_view, failed_branches) =
-        export_changes(mut_repo, &last_export_view, git_repo)?;
-    if let Ok(mut last_export_file) = OpenOptions::new()
-        .write(true)
-        .create(true)
-        .open(&last_export_path)
-    {
-        let thrift_view = simple_op_store_model::View::from(&new_export_store_view);
-        simple_op_store::write_thrift(&thrift_view, &mut last_export_file)
-            .map_err(|err| GitExportError::WriteStateError(err.to_string()))?;
-    }
-    Ok(failed_branches)
-}
-
-fn upgrade_old_export_state(repo: &Arc<ReadonlyRepo>) {
-    // Migrate repos that use the old git_export_operation_id file
-    let last_operation_export_path = repo.repo_path().join("git_export_operation_id");
-    if let Ok(mut last_operation_export_file) = OpenOptions::new()
-        .read(true)
-        .open(&last_operation_export_path)
-    {
-        let mut buf = vec![];
-        last_operation_export_file.read_to_end(&mut buf).unwrap();
-        let last_export_op_id = OperationId::from_hex(String::from_utf8(buf).unwrap().as_str());
-        let loader = repo.loader();
-        let op_store = loader.op_store();
-        let last_export_store_op = op_store.read_operation(&last_export_op_id).unwrap();
-        let last_export_store_view = op_store.read_view(&last_export_store_op.view_id).unwrap();
-        if let Ok(mut last_export_file) = OpenOptions::new()
-            .write(true)
-            .create(true)
-            .open(repo.repo_path().join("git_export_view"))
-        {
-            let thrift_view = simple_op_store_model::View::from(&last_export_store_view);
-            simple_op_store::write_thrift(&thrift_view, &mut last_export_file)
-                .map_err(|err| GitExportError::WriteStateError(err.to_string()))
-                .unwrap();
+    let mut last_export_view = View::new(op_store::View::default());
+    for (git_ref, git_ref_target) in mut_repo.view().git_refs() {
+        if let Some(branch_name) = git_ref.strip_prefix("refs/heads/") {
+            last_export_view.set_local_branch(branch_name.to_string(), git_ref_target.clone());
         }
-        std::fs::remove_file(last_operation_export_path).unwrap();
     }
+    export_changes(mut_repo, &last_export_view, git_repo)
```

Comment:
> I think this is correct even considering e.g. undo. Let's consider
> what would happen in a non-colocated Git repo (not because tricky
> cases cannot happen there but because the explicit exports and imports
> make it easier to discuss, and more cases can occur). If the user
> moved a branch and then did `jj git export`, `jj undo`, and then `jj
> git export` again, we would think on the second export that we should
> perform the same changes to the Git repo, which should have no effect.

I'm not pretty sure if new behavior is more correct. With this PR applied, `jj git export` -> `jj undo` -> `jj branch set` (move) -> `jj git export` would fail because the git state known to jj would be diverged at `jj undo`. Maybe that's the original motivation to record the last exported view/op_id? (I guess the export view needs to be updated on `git import` to snapshot the last known state, though)

OTOH, new behavior is correct given that `jj undo` can't undo any external effects like export refs, push, etc. I think the old behavior is a middle ground to pretend as if `git export` were undo-ed.

Anyway, in colocated environment, git refs would be synced after `jj undo`, so I don't think new behavior makes things worse. And I think a fewer states gives more predictable results.

Reply:
> I'm not pretty sure if new behavior is more correct. With this PR applied, `jj git export` -> `jj undo` -> `jj branch set` (move) -> `jj git export` would fail because the git state known to jj would be diverged at `jj undo`.

True, but at least the failure would be quite benign. The user would need to `jj git import` to get the updated state, resolve any conflicts, and then export again.

> Maybe that's the original motivation to record the last exported view/op_id? (I guess the export view needs to be updated on `git import` to snapshot the last known state, though)

I don't actually remember why I did it that way.

And yes, the export view would need to be updated on import too. That's why that test cases failed in an intermediate commit in this PR. I meant to explain that in the description. I've done that now.

> And I think a fewer states gives more predictable results.

Yes, this is important, IMO. I'm much more confident in the export code now.